src/types.rs

```diff
@@ -335,13 +457,64 @@ pub struct WhereClauseItem {
 /// # struct MyUnitStruct<'a> {
 ///     foo: i32,
 ///     bar: &'a mut Vec<(u64, Option<bool>, f64)>,
-/// };
+/// # };
 /// ```
 #[derive(Clone)]
 pub struct TyExpr {
     pub tokens: Vec<TokenTree>,
 }
 
+impl TyExpr {
+    /// Tries to parse this type in the common form `path::to::Type<'a, other::Type>`.
+    ///
+    /// The 1st element of the tuple is the path, e.g. `path::to::Type`.
+    /// The 2nd element are the raw tokens _inside_ the generic argument list, if available, e.g. `'a, other::Type`.
+    ///
+    /// If this type has any other form, `None` is returned.
+    pub fn as_path(&self) -> Option<(Vec<Ident>, Option<&[TokenTree]>)> {
+        let found_split = self
+            .tokens
+            .iter()
+            .enumerate()
+            .find(|(_, tk)| matches!(tk, TokenTree::Punct(punct) if punct.as_char() == '<'));
+
+        let (prefix, generic_args) = if let Some((split_pos, _)) = found_split {
+            let (prefix, generic_args) = self.tokens.split_at(split_pos);
+
+            let closing_angle_bracket = generic_args.last();
+            if matches!(closing_angle_bracket, Some(TokenTree::Punct(punct)) if punct.as_char() == '>')
+            {
+                (prefix, Some(&generic_args[1..generic_args.len() - 1]))
+            } else {
+                (prefix, None)
+            }
+        } else {
+            // could be 'path::to::Type' without generic args, or something entirely different
+            (self.tokens.as_slice(), None)
+        };
+
+        let path_tokens = tokens_from_slice(prefix);
+        if let Some(path_elems) = try_consume_path(path_tokens) {
+            return Some((path_elems, generic_args));
+        }
+
+        None
+    }
+}
```

Comment:
This whole block should be in `types_edition.rs`.

Reply:
Also, `as_path` isn't a terribly descriptive method name. (though I don't have a better name in mind)

Reply:
Renamed to `interpret_as_path`.

"Path" may sound generic, but [it's the official naming](https://doc.rust-lang.org/reference/paths.html) for token trees like `path::to::Type<'a, other::Type>`.

Reply:
The "as_xxx" naming matches the rest of the crate, though I guess it's not idiomatic rust. My problem is with "path".

> "Path" may sound generic, but [it's the official naming](https://doc.rust-lang.org/reference/paths.html) for token trees like `path::to::Type<'a, other::Type>`.

Yes, but the definition of those paths in the doc doesn't match yours. Eg `mycrate::Foo<Bar>::AssociatedType` is a valid path, though in practice this syntax is rarely used.

Depending on the use case you want for this utility function, this is something to consider.

Reply:
Good point. The general question is also, how far should the parsing go? Path expressions can become arbitrarily complex, e.g.
```rs